### ISSUE TITLE
Erroneous "unresolved symbol" errors when using macros in .cljc files

### ISSUE THREAD

**Issue by simon-katz** [borkdude/clj-kondo#958]
**version**

clj-kondo v2020.07.29

**macro usage**

This bug is related to macro usage, but I'm pretty sure it's not covered by the common options to deal with macros.

**platform**

Native, I think. (I'm on OSX and I've installed clj-kondo using Homebrew.)

**problem**

I've come across a situation where clj-kondo gives erroneous "unresolved symbol" errors when using macros in .cljc files.

I've put some examples in the repository at https://github.com/simon-katz/clj-kondo-cljc-macros

Each of the subdirectories of the `src/clj_kondo_cljs_macros` directory contains an ns.clj file and an ns.cljc file (same namespace name for the two files) that give rise to either correct behaviour or the erroneous report. Each directory also has a README.txt that describes the behaviour.

One of the examples is:

.clj file
=========

```
(ns clj-kondo-cljs-macros.ns2.ns)




(defmacro foo [x y]
  `(when ~x ~y))
```

.cljc file
==========

```
(ns clj-kondo-cljs-macros.ns2.ns
  #?(:cljs
     (:require-macros
      [my-ns :refer [foo]])))

(foo true 1)
```

In the above example, removing one of the blank lines after the ns form in the .clj file or adding a blank line after the ns form in the .cljc file makes the erroneous report go away.

**repro**

https://github.com/simon-katz/clj-kondo-cljc-macros

Output from `clj-kondo --lint src`:

```
src/clj_kondo_cljs_macros/ns2/ns.cljc:6:2: error: unresolved symbol foo
src/clj_kondo_cljs_macros/ns3/ns.cljc:5:2: error: unresolved symbol foo
src/clj_kondo_cljs_macros/ns4/ns.cljc:6:2: error: unresolved symbol foo
linting took 65ms, errors: 3, warnings: 0
```

**config**

The `.clj-kondo/config.edn` file contains an empty map.

**expected behavior**

None of the examples should give errors.

**Comment by borkdude** [comment]
Hey @simon-katz, thanks for your excellent repro.

You are getting these unresolved symbols because clj-kondo lints `.cljc` effectively twice: once for Clojure and once for ClojureScript.

To solve the problem:

```
(ns clj-kondo-cljs-macros.ns2.ns
  #?(:cljs
     (:require-macros
      [my-ns :refer [foo]])))

#?(:cljs (foo true 1)) ;; <- wrap the call inside :cljs only
```

About removing blank lines: that shouldn't matter, so this is definitely a bug in clj-kondo which I will try to fix.

**Comment by simon-katz** [comment]
@borkdude 

Thanks for the quick reply. As you say, wrapping things in `?(:cljs …))` makes the errors go away.

But aren't the calls OK as Clojure code? See the example Clojure REPL session below:

```
$ lein repl
.
.
.
user=> (require 'clj-kondo-cljs-macros.ns2.ns)
nil
user=> (in-ns 'clj-kondo-cljs-macros.ns2.ns)
#namespace[clj-kondo-cljs-macros.ns2.ns]
clj-kondo-cljs-macros.ns2.ns=> (foo true 1)
1
clj-kondo-cljs-macros.ns2.ns=>
```

**Comment by borkdude** [comment]
@simon-katz I think technically you are right, so if there's both a `.clj` and `.cljc` file, clj-kondo should not lint the clojure branches (or non-branched code) of the `.cljc` file? This leads to the question why the file isn't a `.cljs` file instead?

**Comment by simon-katz** [comment]
@borkdude 

The CLJ REPL example in my previous comment seems to show that in CLJ-land the `.clj` and the `.cljc` are both processed when `(require 'clj-kondo-cljs-macros.ns2.ns)` is done. So I think clj-kondo needs to do the same, and that, as now, it should lint all branches.

**Comment by borkdude** [comment]
I don't think so:

``` 
borkdude@MBP2019 /tmp/p1 $ echo "(println :hello)" > src/foo.clj
borkdude@MBP2019 /tmp/p1 $ echo "(println :hello-cljc)" > src/foo.cljc
borkdude@MBP2019 /tmp/p1 $ clj
Clojure 1.10.1
user=> (require 'foo)
:hello
``` 

So the question remains, why is your .cljc file isn't a .cljs file instead, if it's not relevant to Clojure at all?

**Comment by simon-katz** [comment]
@borkdude  Oh! Yes, you are right; if there's a `.clj` file Clojure doesn't bother with the `.cljc` file. (I've been struggling to write code with macros that works with both Clojure and ClojureScript; I thought had found a way, but I haven't.)

So I don't think I'm interested in any changes in this area. As you say, a `.cljs` file would do.

But, FWIW, you asked:

> if there's both a .clj and .cljc file, clj-kondo should not lint the clojure branches (or non-branched code) of the .cljc file?

In that situation, shouldn't non-branched code be linted? (Maybe I'm misunderstanding "non-branched", I'm taking it to mean non-conditional.)

**Comment by borkdude** [comment]
> So I don't think I'm interested in any changes in this area. As you say, a .cljs file would do.

Let's leave it at that. If you have a standalone repro for newlines affecting the amount of errors reported, that would be a bug and I'd be interested in fixing that.

**Comment by rschmukler** [comment]
@borkdude I have run into this behavior and it has lead to needing to add quite a bit of markup. It may be worth considering whether there are some linters (maybe just the unused var linter?) that might benefit from only reporting for cljc if both `cljs` and `clj` modes report the error. Consider the following function:


```clojure
(defn my-side-effecting-fn!
  [common-var ignored-in-cljs-var]
  (let [foo (inc common-var)]
    #?(:clj (side-effect! foo ignored-in-cljs-var))
    (isomorphic-side-effect! foo))
```

This will report an unused variable error in the `cljs` mode.

The alternative then becomes either to have to write:

```clojure
#?(:cljs
   (defn my-side-effecting-fn!
     [common-var _]
     (isomorphic-side-effect! (inc common-var))

   :clj
   (defn my-side-effecting-fn!
     [common-var other-var]
     (let [foo (inc common-var)]
       (side-effect! foo other-var)
      (isomorphic-side-effect! foo))
```

The downside with this approach is that we end up having to maintain two versions of the function.

The alternative then becomes to write something like:

```clojure
(defn my-side-effecting-fn!
  [common-var #?(:cljs _ :clj other-var)]
  (let [foo (inc common-var)]
    #?(:clj (side-effect! common-var other-var))
    (ismorphic-side-effect! foo))
```

This might be the preferred approach, but it does sting the eyes a bit, especially in a more complicated let binding.

Your decision is probably still the right one, but I just wanted to echo that I have actually run into this while writing code that had a subset of side-effecting functions only exist on one side (eg. both sides might log, but only the `clj` side would report metrics.

**Comment by borkdude** [comment]
@rschmukler If `:unused-binding` is bothering you too much, e.g. if you want to turn if off ONLY for Clojure in a `.cljc` file, you can do this:

``` 
(ns clj-kondo-cljs-macros.ns2.ns
  #?(:clj {:clj-kondo/config '{:linters {:unused-binding {:level :off}}}}))

(defn foo [x y]
  #?(:cljs x))
``` 

In that example, only y will be reported unused, for ClojureScript. Both x and y won't be reported for Clojure.

**Comment by rschmukler** [comment]
Nice, that's a very clever workaround! Thank you for the tip.

**Comment by borkdude** [comment]
@rschmukler #920 has been released in the newest version, so you can use that as well.

**Comment by rschmukler** [comment]
Realized I never replied on there, sorry! I am using it, and it's perfect! Thank you for all the work you do!
